File: src/utils/logger.ts
import { invoke } from '@tauri-apps/api/core';

type LogLevel = 'error' | 'warn' | 'info' | 'debug' | 'trace';

class Logger {
  private static instance: Logger;

  private constructor() {}

  static getInstance(): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }
    return Logger.instance;
  }

  private formatMessage(message: string): string {
    const timestamp = new Date().toISOString();
    return `[${timestamp}] ${message}`;
  }

  async log(level: LogLevel, message: string): Promise<void> {
    const formattedMessage = this.formatMessage(message);
    try {
      await invoke('log_message', { level, message: formattedMessage });
    } catch (error) {
      console.error('Failed to log to file:', error);
    }
  }

  async error(message: string): Promise<void> {
    await this.log('error', message);
    console.error(message);
  }

  async warn(message: string): Promise<void> {
    await this.log('warn', message);
    console.warn(message);
  }

  async info(message: string): Promise<void> {
    await this.log('info', message);
    console.log(message);
  }

  async debug(message: string): Promise<void> {
    await this.log('debug', message);
    console.debug(message);
  }

  async trace(message: string): Promise<void> {
    await this.log('trace', message);
    console.trace(message);
  }
}

export const logger = Logger.getInstance();

File: src-tauri/src/logger.rs
use flexi_logger::{Duplicate, FileSpec, Logger, WriteMode};

pub fn init_logger() -> Result<(), Box<dyn std::error::Error>> {
    // Auto-detect debug/release mode
    let log_level = if cfg!(debug_assertions) {
        std::env::var("LOG_LEVEL").unwrap_or_else(|_| "debug".to_string())
    } else {
        std::env::var("LOG_LEVEL").unwrap_or_else(|_| "info".to_string())
    };

    Logger::try_with_str(&log_level)?
        .log_to_file(FileSpec::default()
            .directory("logs")
            .basename("shikimore"))
        .write_mode(WriteMode::Direct)
        .duplicate_to_stderr(Duplicate::Info)
        .start()?;

    Ok(())
}

#[tauri::command]
pub fn log_message(level: String, message: String) {
    match level.as_str() {
        "error" => log::error!("{}", message),
        "warn" => log::warn!("{}", message),
        "info" => log::info!("{}", message),
        "debug" => log::debug!("{}", message),
        "trace" => log::trace!("{}", message),
        _ => log::info!("{}", message),
    }
}
